sdk/servicebus/service-bus/src/core/streamingReceiver.ts

```diff
@@ -338,159 +348,293 @@ export class StreamingReceiver extends MessageReceiver {
+          await userHandlers.processError(args);
+        } catch (err) {
+          await this._reportInternalError(err);
+          logger.logError(err, `An error was thrown from the user's processError handler`);
+        }
+      },
+      processMessage: async (message: ServiceBusMessageImpl) => {
+        try {
+          const span = createProcessingSpan(message, this, this._context.config, operationOptions);
+          return await trace(() => userHandlers.processMessage(message), span);
+        } catch (err) {
+          this._messageHandlers().processError({
+            error: err,
+            errorSource: "processMessageCallback",
+            entityPath: this.entityPath,
+            fullyQualifiedNamespace: this._context.config.host
+          });
           throw err;
         }
+      },
+      postInitialize: async () => {
+        if (!userHandlers.postInitialize) {
+          return;
+        }
 
-        logger.logError(
-          err,
-          `${this.logPrefix} Error thrown in retry cycle ${numRetryCycles}, restarting retry cycle with retry options`,
-          this._retryOptions
+        return userHandlers.postInitialize().catch((err) =>
+          this._messageHandlers().processError({
+            error: err,
+            errorSource: "processMessageCallback",
+            entityPath: this.entityPath,
+            fullyQualifiedNamespace: this._context.config.host
+          })
         );
+      },
+      preInitialize: async () => {
+        if (!userHandlers.preInitialize) {
+          return;
+        }
 
-        continue;
-      }
-    }
+        return userHandlers.preInitialize().catch((err) =>
+          this._messageHandlers().processError({
+            error: err,
+            errorSource: "processMessageCallback",
+            entityPath: this.entityPath,
+            fullyQualifiedNamespace: this._context.config.host
+          })
+        );
+      },
+      forwardInternalErrors: userHandlers.forwardInternalErrors ?? false
+    };
+
+    this._messageHandlers = () => messageHandlers;
   }
 
-  private async _initOnce(args: {
-    useNewName: boolean;
-    abortSignal?: AbortSignalLike;
-  }): Promise<void> {
-    const options = this._createReceiverOptions(args.useNewName, this._getHandlers());
-    await this._init(options, args.abortSignal);
-
-    // this might seem odd but in reality this entire class is like one big function call that
-    // results in a receive(). Once we're being initialized we should consider ourselves the
-    // "owner" of the receiver and that it's now being locked into being the actual receiver.
-    this._receiverHelper.resume();
+  /**
+   * Subscribes using the already assigned `this._messageHandlers` and `this._subscribeOptions`
+   *
+   * @returns A promise that will resolve when a link is created and we successfully add credits to it.
+   */
+  private async _subscribeImpl(caller: "detach" | "subscribe"): Promise<void> {
```

Comment:
The `caller` parameter feels a little odd to me. Are you just trying to get away from using `useNewName`? Maybe something that's a bit more descriptive of what's happening where the values are something like "initialize" and "reconnect" would be a bit clearer without coupling us to the method names calling it?

Reply:
I can change the parameter name but those are the names of the two logical flows I think. 

However, I have thought about renaming 'detach' to 'reconnect' a few times. Maybe this is just a happy chance to do so?